### PR TITLE
Add "Storybook" style parts kit

### PR DESCRIPTION
**Pull request by davist11**
This takes Greg's work and bakes it into the base module with some changes.

- Automatically adds the necessary CSS and JS for storybook style/functionality
- Uses folders instead of special file naming to break apart sections
- Adds a new `getText()` method to randomly generate lorem ipsum (this is used in all the text generator methods)
- Adds the ability to customize the directory, layout, and parts kit volume from a `config/parts-kit.php` file

```twig
{% extends 'viget-base/_layouts/parts-kit' %}

{% block main %}
    <p>{{ craft.viget.partsKit.getText(40) }}</p>

    <p>{{ craft.viget.partsKit.getTitle() }}</p>

    <p>{{ craft.viget.partsKit.getSentence() }}</p>

    <p>{{ craft.viget.partsKit.getParagraph() }}</p>

    {{ craft.viget.partsKit.getRichTextShort() }}

    {{ craft.viget.partsKit.getRichTextFull() }}

    {% set partsKitImage = craft.viget.partsKit.getImage('350x230.png') %}
    {% if partsKitImage %}
        <img src="{{ partsKitImage.url }}">
    {% endif %}
{% endblock %}
```

parts-kit

vigetlabs/craft-starter-module#10

### CODE REVIEW

File: src/resources/css/parts-kit.css [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    fill: #148bbe;
+    margin-right: .25rem;
+    width: .625rem;
+}
+
+.parts-kit__nav-caret {
+    transform: rotate(180deg);
+}
+
+.parts-kit__nav-sub-items[data-togglee] {
```
Comment (olets) [inline src/resources/css/parts-kit.css:95]: spelling? or this "the thing that gets toggled?"

File: src/resources/css/parts-kit.css [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    opacity: 1;
+	transition: all .15s;
+	visibility: visible;
```
Comment (olets) [inline src/resources/css/parts-kit.css]: indentation

File: src/resources/js/parts-kit.js [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    constructor(el) {
+        this.el = el;
+        this.setVars();
+        this.bindEvents();
+        this.initSidebar();
+    }
+
+    setVars() {
+        this.sidebar = this.el.querySelector("[data-sidebar]");
+        this.lsSidebar = "pk-sidebar-hidden";
```
Comment (olets) [inline src/resources/js/parts-kit.js:23]: what's `ls`?

File: src/resources/js/parts-kit.js [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        this.el.addEventListener("transitionend", this.resetHeight, false);
+        this.el.style.maxHeight = this.el.scrollHeight + "px";
+        this.el.style.overflow = "hidden";
+
+        // Wait a tick so height has applied
+        requestAnimationFrame(() => {
+            requestAnimationFrame(() => {
+                this.el.classList[method](this.collapsedClass);
+            });
+        });
```
Comment (olets) [inline src/resources/js/parts-kit.js:94]: 🧠

File: src/Bundle.php [hunk 5]
```diff
@@ -2,18 +2,31 @@
...
 
-        $this->css = [
-            'style.css',
+        $css = [
+            'css/style.css'
         ];
 
+        // Is this a parts kit request
+        if (($firstSegment = Craft::$app->request->segments[0] ?? null) === PartsKit::getConfig('directory')) {
+            $css[] = 'css/parts-kit.css';
```
Comment (gregkohn) [inline src/Bundle.php:21]: love this extendable approach

File: src/Module.php [hunk 6]
```diff
@@ -126,12 +125,23 @@ class="edit-entry"
                 }
             }
         );
+
+        // Define template directory for site
+        if (($firstSegment = Craft::$app->request->segments[0] ?? null) === PartsKit::getConfig('directory')) {
```
Comment (gregkohn) [inline src/Module.php]: both here and in `Bundle.php`...why is this set to `$firstSegment`? considering you do this logic twice, how would you feel about a parts kit service method like `PartsKit::isRequest` or something?

File: src/services/PartsKit.php [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+            ];
+
+            foreach ($files as $file) {
+                $url = UrlHelper::siteUrl(implode('/', [
+                    $partsKitDir,
+                    $dir,
+                    $file,
+                ]));
+
+                $isActive = $url === Craft::$app->request->absoluteUrl;
```
Comment (gregkohn) [inline src/services/PartsKit.php:80]: 💯  nice, my version was flawed

**Review by gregkohn — APPROVED** [review]
🥇  lmk when i should try dropping this into Trust

File: src/resources/js/parts-kit.js [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+    };
+}
+
+// Init the parts kit class
+const partsKitEl = document.querySelector("[data-parts-kit]");
+new PartsKit(partsKitEl);
+
+// Slide class for animating hiding/showing
+class Slide {
+    constructor(el, collapsedClass = "collapsed") {
```
Comment (olets) [inline src/resources/js/parts-kit.js:66]: Is support for a custom `collapsedClass` needed/used?

File: src/templates/_layouts/parts-kit.html [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+                                    ) }}
+                                    {{ svg(
+                                        '@viget/base/resources/icons/pk-component.svg',
+                                        class='parts-kit__nav-component'
+                                    ) }}
+                                    <h3>{{ title | title }}</h3>
+                                </div>
+                            </button>
+
+                            <ol
```
Comment (olets) [inline src/templates/_layouts/parts-kit.html:37]: My understanding is for a11y this should have an `aria-labelledby` pointing back to the `button`

File: src/resources/js/parts-kit.js [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+            onClose: null,
+            onToggle: null,
+        };
+
+        this.setVars();
+        this.bindEvents();
+    }
+
+    setVars() {
+        this.togglers = this.el.querySelectorAll("[aria-controls]");
```
Comment (olets) [inline src/resources/js/parts-kit.js:125]: Looks like there's only ever one?

File: src/resources/js/parts-kit.js [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+        this.callback("onToggle");
+    };
+
+    close = () => {
+        this.isOpen = false;
+        this.togglee.close();
+        this.togglers.forEach((el) =>
+            el.setAttribute("aria-expanded", "false")
+        );
+        this.el.removeAttribute("open");
```
Comment (olets) [inline src/resources/js/parts-kit.js:159]: Is this attribute used?

File: src/resources/js/parts-kit.js [hunk 12]
```diff
@@ -0,0 +1,186 @@
+const keyCodes = {
+    tab: 9,
+    enter: 13,
+    esc: 27,
+    arrowLeft: 37,
+    arrowUp: 38,
+    arrowRight: 39,
+    arrowDown: 40,
+    f: 70,
+};
```
Comment (olets) [inline src/resources/js/parts-kit.js:10]: Only esc and f are used. Would be cool to follow the full keyboard recs https://www.w3.org/TR/wai-aria-practices-1.1/examples/accordion/accordion.html but that might be a future PR. In the meantime do we need these other entries?

**Review by olets — APPROVED** [review]
This is really exciting. Can't wait to put it on a site and play with it.

File: src/resources/css/parts-kit.css (hunk 1 above)
Comment (olets) [reply src/resources/css/parts-kit.css:95]: nevermind, answered my own question by reading the rest.

File: src/resources/js/parts-kit.js (hunk 8 above)
Comment (gregkohn) [reply src/resources/js/parts-kit.js:66]: yeah, some of these things came from some actual utilities i had on Trust so there's a bit of flexibility that doesn't actually need to be there in this locked-down version

Comment (olets) [reply src/resources/js/parts-kit.js:66]: got it that makes sense

File: src/Module.php (hunk 6 above)
Comment (davist11) [reply src/Module.php]: Ahh guess i don't need to do it like this now. But basically that means you can use `$firstSegment` inside of the condition.

But yeah, I like your idea of having a static method

File: src/resources/js/parts-kit.js (hunk 12 above)
Comment (davist11) [reply src/resources/js/parts-kit.js:10]: Yeah this was all copied pretty verbatim from @gregkohn's work, so on the fence about how much time we spend cleaning it up vs keeping it extendable.

File: src/resources/js/parts-kit.js (hunk 3 above)
Comment (davist11) [reply src/resources/js/parts-kit.js:23]: 🤷

Comment (gregkohn) [reply src/resources/js/parts-kit.js:23]: woops, it's for `localStorage`. clearly a bad name, im sure there's a better convention for storing that entry name

**Pull request merged**